### PR TITLE
Feature/rest api allow config updates

### PR DESCRIPTION
**Pull request by OwenTuz**
I believe this is a working branch with tests, but there are a few things worth discussing - I'll add notes to the commits.

### CODE REVIEW

**Comment by andresriancho** [comment]
Awesome PR! I believe it's one of the most complex PRs w3af has ever received. Starting my review.

File: w3af/core/controllers/core_helpers/status.py [hunk 1]
```diff
@@ -166,13 +166,13 @@ def get_rpm(self):
                  start of the scan.
         """
         if self._start_time_epoch is None:
-            raise RuntimeError('Can NOT call get_run_time before start().')
-        
+            return None
```
Comment (andresriancho) [inline w3af/core/controllers/core_helpers/status.py]: Hmmm, why this change? I believe that the exception was correct, but if you've got a good reason we can keep it.

File: w3af/core/ui/api/resources/scans.py [hunk 2]
```diff
@@ -64,9 +64,14 @@ def start_scan():
     # REST API is an MVP and we can only run one scan at the time (for now)
     #
     scan_infos = SCANS.values()
-    if not all([si is None for si in scan_infos]):
+    if not all([
+        (
+        si is None or
+        si.w3af_core.status.get_simplified.status.upper == 'STOPPED'
```
Comment (andresriancho) [inline w3af/core/ui/api/resources/scans.py]: Should be `get_simplified_status().upper()` ?

File: w3af/core/ui/api/resources/sessions.py [hunk 3]
```diff
@@ -0,0 +1,343 @@
...
+                                             enable_or_disable_plugin)
+from w3af.core.ui.api.db.master import SCANS
+from w3af.core.ui.api.utils.scans import (get_scan_info_from_id,
+                                          create_scan_helper,
+                                          start_preconfigured_scan)
+from w3af.core.controllers.w3afCore import w3afCore
+from w3af.core.controllers.exceptions import BaseFrameworkException
+from w3af.core.controllers.misc_settings import MiscSettings
+from w3af.core.data.options.opt_factory import opt_factory
+from w3af.core.data.options.option_types import *
```
Comment (andresriancho) [inline w3af/core/ui/api/resources/sessions.py]: Please avoid wildcard imports

File: w3af/core/ui/api/resources/sessions.py [hunk 4]
```diff
@@ -0,0 +1,343 @@
...
+        - The newly created scan ID (eg. 0)
+    """
+    scan_id = create_scan_helper()
+
+    return jsonify({'message': 'Success',
+                    'id': scan_id,
+                    'href': '/sessions/%s' % scan_id}), 201
+
+
+@app.route('/sessions/<int:scan_id>', methods=['GET'])
```
Comment (andresriancho) [inline w3af/core/ui/api/resources/sessions.py]: Just a naming thing but... `int:scan_id` should be `int:session_id`?

File: w3af/core/ui/api/resources/sessions.py [hunk 5]
```diff
@@ -0,0 +1,343 @@
...
+
+    :return: 404 if scan does not exist.
+             Otherwise, a list of possible routes under /sessions/<scan_id>/
+    """
+    urls = list_subroutes(request)
+    return jsonify({'message': 'Session %s' % scan_id,
+                    'available_endpoints': urls})
+
+
+@app.route('/sessions/<int:scan_id>/plugins/', methods=['GET'])
```
Comment (andresriancho) [inline w3af/core/ui/api/resources/sessions.py]: Same as above with `int:session_id`

Comment (andresriancho) [inline w3af/core/ui/api/resources/sessions.py]: This repeats in almost all endpoints, please fix in all.

File: w3af/core/ui/api/resources/sessions.py [hunk 6]
```diff
@@ -0,0 +1,343 @@
...
+            opt_type = plugin_opts[opt_name].get_type()
+        except BaseFrameworkException:
+            return jsonify({
+                'code': '400',
+                'message': '%s is not a valid option for plugin %s' %
+                    (opt_name,
+                     plugin)
+                }), 400
+        try:
+            if opt_type.lower() == 'url_list':
```
Comment (andresriancho) [inline w3af/core/ui/api/resources/sessions.py]: I believe that this is not required, if you do a `plugin_opts[opt_name].set_value(['http://x.com', 'http://y.com'])` it should work.

File: w3af/core/ui/api/resources/sessions.py [hunk 7]
```diff
@@ -0,0 +1,343 @@
...
+            else:
+                plugin_opts[opt_name].set_value(opt_value)
+        except (AttributeError, BaseFrameworkException):
+            return jsonify({
+                'code': '422',
+                'message': 'Invalid %s option %s' % (opt_type, opt_value)
+                }), 422
+
+    with SET_OPTIONS_LOCK:
+        if 'enabled' in request.json:
```
Comment (andresriancho) [inline w3af/core/ui/api/resources/sessions.py]: Didn't you handle the `enabled` parameter before?

File: w3af/core/ui/api/resources/sessions.py [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+                                     plugin,
+                                     plugin_type,
+                                     enable=request.json['enabled'])
+            return jsonify({
+                'message': 'success',
+                'modified': request.json
+                })
+    try:
+        # Remove 'enabled' if included as we must set this separately
+        opt_names.pop('enabled')
```
Comment (andresriancho) [inline w3af/core/ui/api/resources/sessions.py]: We're going to have a bug here when a plugin (for some strange reason) adds a configuration parameter name called `enabled`.

Maybe you can add code to `OptionList` to make sure that never happens? I don't mind restricting the plugin developers like that.

Another solution is to have the `enabled` and plugin-related options in different JSON "namespaces".

File: w3af/core/ui/api/resources/sessions.py [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+        try:
+            opt_value = request.json[opt_name]
+            opt_type = core_opts[opt_name].get_type()
+        except BaseFrameworkException:
+            return jsonify({
+                'code': '400',
+                'message': '%s is not a valid option here' % opt_name
+                }), 400
+        try:
+            if opt_type.lower() == 'url_list':
```
Comment (andresriancho) [inline w3af/core/ui/api/resources/sessions.py]: Same as above

File: w3af/core/ui/api/resources/sessions.py [hunk 10]
```diff
@@ -0,0 +1,343 @@
...
+    w3af = SCANS[scan_id].w3af_core
+
+    if core_setting == 'http':
+        configurable = w3af.uri_opener.settings
+    if core_setting == 'misc':
+        configurable = MiscSettings()
+    if core_setting == 'target':
+        configurable = w3af.target
+
+    core_opts = configurable.get_options()
```
Comment (andresriancho) [inline w3af/core/ui/api/resources/sessions.py]: The next ~30 lines look very much like the configuration of plugin parameters. Maybe you should avoid code duplication and move this to a helper?

File: w3af/core/ui/api/tests/test_session_scan.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+class APIScanFromSessionTest(IntegrationTest):
+
+    def test_start_scan_from_session(self):
+        plugin_settings = {
+            "crawl.web_spider": {
+                "only_forward": True,
+                "follow_regex": ".*"},
+            "audit.sqli": {}
+        }
+        target_url = "http://127.0.0.1:8000/audit/sql_injection/"
```
Comment (andresriancho) [inline w3af/core/ui/api/tests/test_session_scan.py]: Use `get_moth_http` (or similar function, see other tests) to avoid hard-coding the port

File: w3af/core/ui/api/tests/test_session_scan.py [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+                    self.api_url, scan_id, cs),
+                auth = self.api_auth,
+                data=json.dumps(core_settings[cs]),
+                headers=self.headers)
+
+        start_scan = requests.post(
+            '%s/sessions/%s/start' % (
+                self.api_url, scan_id),
+            auth = self.api_auth,
+            headers=self.headers)
```
Comment (andresriancho) [inline w3af/core/ui/api/tests/test_session_scan.py]: Test has really good coverage, well done :+1:

File: w3af/core/ui/api/tests/test_sessions.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+        session_id = json.loads(create_session.data)['id']
+        self.assertEqual(create_session.status_code, 201)
+
+        get_plugin_types = self.app.get('/sessions/%s/plugins/' % session_id,
+                                       headers=self.HEADERS)
+        plugin_types = json.loads(get_plugin_types.data)['entries']
+        self.assertEqual(get_plugin_types.status_code, 200)
+        self.assertIsInstance(plugin_types, list)
+        self.assertTrue('audit' in plugin_types)
```
Comment (andresriancho) [inline w3af/core/ui/api/tests/test_sessions.py]: `self.assertIn` should be used in these cases

File: w3af/core/ui/api/tests/test_sessions.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+            '/sessions/%s/plugins/audit/this_should_also_404/' % session_id,
+            '/sessions/%s/core/nonexistent_setting/' % session_id
+            ]
+
+        for endpoint in test_404s:
+            should_404 = self.app.get(endpoint,
+                                       headers=self.HEADERS)
+            code = json.loads(should_404.data)['code']
+            self.assertEqual(code, 404)
+            self.assertEqual(should_404.status_code, 404)
```
Comment (andresriancho) [inline w3af/core/ui/api/tests/test_sessions.py]: This is a really long test that looks like it could be split in various tests. It's always better to have isolated tests which assert specific things.

**Comment by andresriancho** [comment]
PR review finished! The code looks good, there are tests which PASS, everything seems to be complete except from :skull: :skull: :skull: documentation. Please consider adding documentation for the `sessions` endpoint to https://github.com/andresriancho/w3af/tree/master/doc/sphinx/api , follow the formatting/wording from the other endpoints. Documentation is critical since we want people to consume the API, and without docs they will find it really hard.

Add the docs to this PR, I'll review them, run the tests and code and finally merge.

File: w3af/core/ui/api/resources/sessions.py (hunk 6 above)
Comment (OwenTuz) [inline w3af/core/ui/api/resources/sessions.py]: I receive an AttributeError if I do this - it looks like the code in set_value is expecting a URL object and calling verify_url:
`{
  "code": 500, 
  "exception_type": "AttributeError", 
  "filename": "target.py", 
  "function_name": "_verify_url", 
  "line_number": 101, 
  "message": "'unicode' object has no attribute 'get_protocol'", 
  "please": "https://github.com/andresriancho/w3af/issues/new"
}`

Presumably the user could call URL() and pass a URL object in that list with some success, but it seemed less friendly than plain text: there may be a better way to handle this?

File: w3af/core/ui/api/resources/sessions.py (hunk 7 above)
Comment (OwenTuz) [inline w3af/core/ui/api/resources/sessions.py]: I've added a comment to clarify this:

`     if ('enabled' in opt_names and len(opt_names) == 1):
+    # If the request only contains 'enabled', set it here, return and skip
+    # the additional logic.
+    # If the request contains 'enabled' alongside other options, we handle it
+    # in the same lock section as other options below.
+    #
+    # This is done to ensure that the PATCH request is always atomic.
         with SET_OPTIONS_LOCK:
             enable_or_disable_plugin(w3af,`

I think it's valid, as the PATCH request must be atomic, but perhaps there's a neater way to handle it?

File: w3af/core/ui/api/resources/sessions.py (hunk 8 above)
Comment (OwenTuz) [inline w3af/core/ui/api/resources/sessions.py]: Good catch, thanks. I'm afraid I've taken the more restrictive path and caught this in the 'add' method for `OptionList` - while the namespace method seems more elegant, it seems as though it would be confusing to allow 'enabled' as a plugin option.

**Comment by OwenTuz** [comment]
Thanks for taking the time to review it, and for the feedback - I've implemented some of these and commented on others. 

Still to do is refactoring to reduce code reuse, so I'll add that as soon as I get a chance, along with any other fixes/edits based on the above.

One note - I'm finding that get_moth_http() returns 'fallback' and breaks the scan test, but it does so for existing tests as well. I'm getting to grips with a Mac for the first time due to work VPN requirements (Citrix >8( long story) and suspect this is my environment rather than the code.

If it does fail your CI tests, though, I apologise - let me know and I'll figure out an alternative.

**Comment by OwenTuz** [comment]
Also, yes. The documentation :disappointed:

File: w3af/core/controllers/core_helpers/status.py (hunk 1 above)
Comment (OwenTuz) [inline w3af/core/controllers/core_helpers/status.py]: This is called from `/scans/<scan_id>/status`, via `w3af.core.status.get_status_as_dict()`
The dict looks like this:
```
 data = {                                                        
         'status': self.get_simplified_status(),                 
         'is_paused': self.is_paused(),                          
         'is_running': self.is_running(),                        
                                                                 
         'active_plugin':                                        
             {'crawl': self.get_running_plugin('crawl'),         
              'audit': self.get_running_plugin('audit')}         
         ,                                                       
                                                                 
         'current_request':                                      
             {'crawl': crawl_fuzzable_request,                   
              'audit': audit_fuzzable_request},                  
                                                                 
         'queues':                                               
             {'crawl':                                           
                 {'input_speed': self.get_crawl_input_speed(),   
                  'output_speed': self.get_crawl_output_speed(), 
                  'length': self.get_crawl_qsize()},             
              'audit':                                           
                 {'input_speed': self.get_audit_input_speed(),   
                  'output_speed': self.get_audit_output_speed(), 
                  'length': self.get_audit_qsize()}              
             },                                                  
                                                                 
         'eta':                                                  
             {'crawl': self.get_crawl_eta(),                     
              'audit': self.get_audit_eta()},                    
                                                                 
         'rpm': self.get_rpm(),                                  
         }                                                       
```

Of those, only `get_rpm()` was throwing an exception if called before the scan was started - the rest return 'None', though functions such as `get_crawl_eta()` are equally meaningless if the scan has never run.

It may be more correct to patch the other functions so that they throw an exception, and simply handle the exceptions in `/scans/<scan_id>/status` instead?

**Comment by OwenTuz** [comment]
Updated, with notes re some of your comments - there may well be better ways to do this, so please let me know if you have any preference.

**Comment by andresriancho** [comment]
Going to work on this PR today, I'll test manually, review the source and merge. These changes are really valuable and should be in develop/master
